core/server/master/src/main/java/alluxio/master/thrift/SocketTrackingTServerSocket.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.thrift;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+
+import com.google.common.io.Closer;
+import org.apache.thrift.transport.TServerSocket;
+import org.apache.thrift.transport.TSocket;
+import org.apache.thrift.transport.TTransportException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Extension of TServerSocket which tracks all accepted sockets and closes them when the server
+ * socket is closed.
+ */
+public class SocketTrackingTServerSocket extends TServerSocket {
+  private static final Logger LOG = LoggerFactory.getLogger(SocketTrackingTServerSocket.class);
+  private static final long CLEANUP_INTERVAL_MS =
+      Configuration.getMs(PropertyKey.MASTER_CLIENT_SOCKET_CLEANUP_INTERVAL);
+
+  private final Set<Socket> mSockets = ConcurrentHashMap.newKeySet();
```

Comment:
Does `KeySetView` work the way we want, or should we use `ConcurrentHashSet`?

Reply:
It works